### ISSUE TITLE
Validators.required will return error if object has a property named length.

### ISSUE THREAD

**Issue by mtricolici98** [angular/angular#36858]
**Description:**
When validating an object that has property `length`, validator will return an error if length value is 0. 

Traced the problem all the way down to this code line.
https://github.com/angular/angular/blob/89c589085d89205223c440f003dbea43c6523928/packages/forms/src/validators.ts#L18

**Comment by danranVm** [comment]
```ts
   return value == null || value.length === 0 && (typeof obj === 'string' || Array.isArray(obj));
```
@AndrewKushnir Is this right? you can assign it to me, i will fix it and add test.

**Comment by sonukapoor** [comment]
> **Description:**
> When validating an object that has property `length`, validator will return an error if length value is 0.
> 

@mtricolici98 Are you seeing this issue because you are using the `minLength` or `maxLength` validators?

**Comment by mtricolici98** [comment]
@sonukapoor , No, the issue is when using the "Required" validator, and the object has a property called `length`. 

@danranVm, That might do the trick, I thought maybe just checking if the value is not an object would suffice though. Although I might not be aware of corner cases.

**Comment by is2ei** [comment]
@sonukapoor  (Cc: @mtricolici98 )

I created reproduction example:
https://stackblitz.com/edit/angular-c1qmmm?file=src%2Fapp%2Fapp.component.ts

```typescript
    const value: {} = { length: 0 }; // The validator says this is empty. But it's not correct.
    // const value: {} = { length: 1 }; // The validator says this is not empty. It's correct.
    this.testForm = new FormGroup({
      'testInput': new FormControl(value, [
        Validators.required
      ])
    });
```

**Comment by AndrewKushnir** [comment]
@mtricolici98, thanks for reporting the issue.

@danranVm, the direction of the change that you proposed looks correct, but we need to review all validators to see how they currently work with objects that have `length` property (create a few tests to illustrate the behavior). Once we have this coverage, we can look into changing the behavior and verify whether the change is breaking or not.

**Comment by JeanMeche** [comment]
This has been fixed by #33729 and released in v14.
